src/strtree.c

```diff
@@ -553,6 +555,489 @@ static PyObject* STRtree_query_bulk(STRtreeObject* self, PyObject* args) {
+  get_geom(query_pg_geom, &query_geom);
+
+  tree_nearest_userdata_t* params = (tree_nearest_userdata_t*)userdata;
+
+  // distance returns 1 on success, 0 on error
+  if (GEOSDistance_r(params->ctx, query_geom, tree_geom, &calc_distance) == 0) {
+    return 0;
+  }
+
+  // store any pairs that are smaller than the minimum distance observed so far
+  // and update min_distance
+  if (calc_distance <= params->min_distance) {
+    params->min_distance = calc_distance;
+
+    // if smaller than last item in vector, remove that item first
+    pairs_size = kv_size(*(params->dist_pairs));
+    if (pairs_size > 0 &&
+        calc_distance < (kv_A(*(params->dist_pairs), pairs_size - 1)).distance) {
+      kv_pop(*(params->dist_pairs));
+    }
+
+    tree_geom_dist_vec_item_t dist_pair =
+        (tree_geom_dist_vec_item_t){(GeometryObject**)item1, calc_distance};
+    kv_push(tree_geom_dist_vec_item_t, *(params->dist_pairs), dist_pair);
+  }
+
+  // Set distance for callback
+  // This adds a slight adjustment to force checking of adjacent tree nodes; otherwise
+  // they are skipped by the GEOS nearest neighbor algorithm check against bounds
+  // of adjacent nodes.
+  *distance = calc_distance + 1e-6;
+
+  return 1;
+}
+
+#if GEOS_SINCE_3_6_0
+
+/* Find the nearest singular item in the tree to each input geometry.
+ * Returns indices of source array and tree items.
+ *
+ * If there are multiple equidistant or intersected items, only one is returned,
+ * based on whichever nearest tree item is visited first by GEOS.
+ *
+ * Returns a tuple of empty arrays of shape (2,0) if tree is empty.
+ *
+ *
+ * Returns
+ * -------
+ * ndarray of shape (2, n) with input indexes, tree indexes
+ * */
+
+static PyObject* STRtree_nearest(STRtreeObject* self, PyObject* arr) {
+  PyArrayObject* pg_geoms;
+  GeometryObject* pg_geom = NULL;
+  GEOSGeometry* geom = NULL;
+  GeometryObject** nearest_result = NULL;
+  npy_intp i, n, size, geom_index;
+  char* head_ptr = (char*)self->_geoms;
+  PyArrayObject* result;
+
+  // Indices of input and tree geometries
+  index_vec_t src_indexes;
+  index_vec_t nearest_indexes;
+
+  if (self->ptr == NULL) {
+    PyErr_SetString(PyExc_RuntimeError, "Tree is uninitialized");
+    return NULL;
+  }
+
+  if (!PyArray_Check(arr)) {
+    PyErr_SetString(PyExc_TypeError, "Not an ndarray");
+    return NULL;
+  }
+
+  pg_geoms = (PyArrayObject*)arr;
+  if (!PyArray_ISOBJECT(pg_geoms)) {
+    PyErr_SetString(PyExc_TypeError, "Array should be of object dtype");
+    return NULL;
+  }
+
+  if (PyArray_NDIM(pg_geoms) != 1) {
+    PyErr_SetString(PyExc_TypeError, "Array should be one dimensional");
+    return NULL;
+  }
+
+  // If tree is empty, return empty arrays
+  if (self->count == 0) {
+    npy_intp index_dims[2] = {2, 0};
+    result = (PyArrayObject*)PyArray_SimpleNew(2, index_dims, NPY_INTP);
+    return (PyObject*)result;
+  }
+
+  n = PyArray_SIZE(pg_geoms);
+
+  // preallocate arrays to size of input array
+  kv_init(src_indexes);
```

Comment:
That doesn't work, due to missing / empty input geometries: those will never have a nearest geometry in the tree.  So we will have no more than the number of input items, but we may have less.

Alternatively, we could return `nan` for result indices, but that seems awkward because it would be a different data type than `np.intp`.

Reply:
Right, I agree with that. Let's keep it like this then :+1: